### PR TITLE
feat(all): Configure plugins with AmplifyOutputs

### PR DESCRIPTION
**Pull request by lawmicha**
## Description
This PR adds a new configuration type called **AmplifyOutputsData** under `@_spi`. The entry point for customers to configure Amplify with this new configuration is with the **AmplifyOutputs** type.

**Using `amplify_outputs.json` file**.
The file is generated from Amplify CLI Gen2 as `amplify_outputs.json`. An example of the steps to get this file in each respective integration test target's README.md. Developers can add this file to the app bundle and call:

```swift
try Amplify.configure(with: .amplifyOutputs)
```

**Using a resource from the Bundle**

A renamed version of the file, such as "amplify-outputs-dev.json" added to the main Bundle can be accessed by the name of the resource:

```swift
try Amplify.configure(with: .resource(named: "amplify_outputs_dev"))
```

**Using Data object**

The file read into memory and can be passed to Amplify, via:

```swift
let data = /* Read amplify-outputs.json file into memory */
try Amplify.configure(with: .data(data))
```

## API

**AWSAPIPlugin** supports mapping the data category from AmplifyOutputs into a single GraphQL endpoint during configuration. 

## Analytics
**AWSPinpointAnalyticsPlugin** supports mapping the analytics category from AmplifyOutputs to retrieve the appId and region. 

Analytics contain manually modified configuration values from `amplifyconfiguration.json`. As we want to avoid putting library-specific configuration in the backend-specific configuration file, we will provide an Options object that can be passed in when the plugin is instantiated:

```swift
// For context to show available options properties, not complete code snippet
extension AWSPinpointAnalyticsPlugin {
   public struct Options {
       // ... defaults 
       public let autoFlushEventsInerval: TimeInterval
       public let trackAppSessions: Bool
    // ..
   }
}

// Developer call site:
let plugin = AWSAnalyticsPlugin(options: .init(autoFlushEventsInterval: 10))
Amplify.addPlugin(plugin)
Amplify.configure(with: .amplifyOutputs)
```

## Auth

**AWSCognitoAuthPlugin** supports reading the auth category section from AmplifyOutputs. This is mostly mapping with some defaults, specific call outs are

- authFlowType to `.userSRP`. 
- some features are currently not supported, such as custom domain, setting a pinpoint app id for auth events, or the clientSecret.
- use the first redirect signIn/signOut URL available in the array from AmplifyOutputs
- oauth section in AmplifyOutputs contains optional `customDomain` which overrides the `cognitoDomain`.

The additional properties added to **UserPoolConfigurationData**  represents the properties that will be used by the Authenticator. The types such as **PasswordProtectionSettings**, **UsernameAttribute**, **SignUpAttributeType**, and **VerificationMechanism** that the Authenticator have defined have been moved over to this plugin. 
1. We parse the AmplifyOutputs into these migrated types from Authenticator under the UserPoolConfigurationData. 
2. Since Authenticator depends on the JSONValue representation of `amplifyconfiguration.json`, we will then translate the `UserPoolConfigurationData` sections that are relevant for Authenticator back to the `JSONValue` object that it is expecting. This is required for Gen2 usage since Authenticator does not pin it's dependency of Amplify. It will continue to read and parse the `JSONValue` object. 
3. After this PR, we will create a change in Authenticator to the released changes and begin reading from a type instead of parsing the `JSONValue`. Then in the plugin, we can remove the translation from the type back to the JSONValue.

## DataStore

No changes have been made *AWSDataStorePlugin**. It will receive either the AmplifyConfiguration JSONValue or the AmplifyOutputsData when configure is called, and it doesn't validate or use it in any way.

## Geo

**AWSLocationGeoPlugin** supports reading the geo category section from AmplifyOutputs. We do the same type of logic as before, in addition to mapping it back to the internal types, we include the same validation logic as before, such as making sure that the `defaultMap` exists in the `maps` and `defaultSearchIndex` exists in the `searchIndices`

## Logging

No changes have been made to **AWSCloudWatchLoggingPlugin**, it's configure method uses the `amplifyconfiguration_logging.json` configuration, or can be configured programmatically when the plugin is instantiated. There is no Gen2 category for logging. 

## Notifications

**AWSPinpointPushNotificationsPlugin** supports reading from the notifications category section of AmplifyOutputs. This basically retrieves the appId and region for pinpoint.

## Predictions
**AWSPredictionsPlugin** currently does not have category configuration in AmplifyOutputs so the library will throw if the predictions plugin was added for Gen2 customers. 

## Storage
**AWSS3StoragePlugin** supports reading from the storage category of AmplifyOutputs. There is no longer `defaultAccessLevel` configuration and we do not plan to add this as we look to deprecate **StorageAccessLevel** for the `path` based APIs.

## Integration testing

For each category, we have created a new schema that mirrors the old, like "AWSS3StoragePlugin**Gen2**IntegrationTests", with a separate test plan. The test plan allows us to incrementally enable tests that run against the Gen2 baackend. The instructions for running these tests are added as a section to the existing README.md of each respective integration test target. 

An example of the README.md https://github.com/aws-amplify/amplify-swift/pull/3567/files#diff-27ccacf97b7541451f1082b6b5235253fe348e5e5fab37c91be2b6c9d2994b59

The README.md contains similar steps as previously, except instead of running Amplify CLI commands, we define the usage of the Amplify CDK constructs in code using Amplify CLI gen2. The output file is copied over to the test directory and the test code has been updated to use the Gen2 config file `amplify_outputs.json` when running the Gen2 test schema.

## General Checklist

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [x] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [x] Documentation update for the change if required
- [x] PR title conforms to conventional commit style
- [ ] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/CodeGen/Data/UserPoolConfigurationData.swift [hunk 1]
```diff
@@ -83,3 +100,136 @@ extension UserPoolConfigurationData.CustomEndpoint {
         validatedHost = endpoint.host
     }
 }
+
+extension UserPoolConfigurationData {
+
+    /// settings used in the Authenticator
+    struct PasswordProtectionSettings: Equatable, Codable {
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/CodeGen/Data/UserPoolConfigurationData.swift:107]: In the current PR, we will translate the object back to the `@_spi` backed `jsonConfiguration` property on the plugin that Authenticator uses https://github.com/aws-amplify/amplify-swift/blob/main/AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/AWSCognitoAuthPlugin.swift#L39

To answer your question, Authenticator will later use these types instead of the `JSONValue`, but when we do this, we will still use `@_spi`, so these types won't be public

[the remark this answers is not in the document]
Comment (ruisebas) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/StateMachine/CodeGen/Data/UserPoolConfigurationData.swift:107]: Aah my bad, I thought we were putting the migration into `AuthConfiguration` into this work as well.
But got it, the Authenticator will keep using `jsonConfiguration` for now, which I see is being manually built to look like the old one :) 👍

File: AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        public static let defaultTrackAppSession = true
+        public static let defaultAutoSessionTrackingInterval: UInt = {
+        #if os(macOS)
+            .max
+        #else
+            5
+        #endif
+        }()
+
+        public let autoFlushEventsInterval: UInt
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift:22]: Yes, the intention was to move the internal `Int` to `UInt` so it can't be negative. The API discussion was geared towards where we want to place the "options" type, without much focus on the property naming and types themselves. Will look at this further

[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift]: Updated PR description and PR to use TimeInterval

File: Amplify/Core/Configuration/Internal/AmplifyConfigurationInitialization.swift [hunk 3]
```diff
@@ -73,3 +73,74 @@ extension AmplifyConfiguration {
...
+        let jsonDecoder = JSONDecoder()
+
+        do {
+            jsonDecoder.keyDecodingStrategy = .convertFromSnakeCase
+            let configuration = try jsonDecoder.decode(AmplifyOutputsData.self, from: data)
+            return configuration
+        } catch {
+            throw ConfigurationError.unableToDecode(
+                """
+                Could not decode `amplify_outputs.json` into a valid AmplifyConfigurationV2 object
```
Comment (thisisabhash) [inline Amplify/Core/Configuration/Internal/AmplifyConfigurationInitialization.swift:133]: nit: should be - *into a AmplifyOutputsData object*

File: Amplify/Core/Configuration/AmplifyOutputsData.swift [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    /// **Lifecycle**
+    ///
+    /// Internally, Amplify configures the Hub and Logging categories first, so they are available to plugins in the
+    /// remaining categories during the configuration phase. Plugins for the Hub and Logging categories must not
+    /// assume that any other categories are available.
+    ///
+    /// After Amplify has configured all of its categories, it will dispatch a `HubPayload.EventName.Amplify.configured`
+    /// event to each Amplify Hub channel. After this point, plugins may invoke calls on other Amplify categories.
+    ///
+    /// - Parameter configuration: The AmplifyConfigurationV2 for specified Categories
```
Comment (thisisabhash) [inline Amplify/Core/Configuration/AmplifyOutputsData.swift:299]: AmplifyConfigurationV2 -> AmplifyOutputsData

**Review by thisisabhash — COMMENTED** [review]
The Amplify/Core and AmplifyPlugins/Geo folders look good. Added a few minor comments. Will approve when other plugins are reviewed too.

File: AmplifyPlugins/API/Sources/AWSAPIPlugin/Configuration/AWSAPICategoryPluginConfiguration.swift [hunk 5]
```diff
@@ -160,6 +180,25 @@ public struct AWSAPICategoryPluginConfiguration {
...
+        if let data = config.data {
+            let name = "dataCategory"
+            let endpointConfig = try EndpointConfig(name: name,
+                                                    config: data,
+                                                    apiAuthProviderFactory: apiAuthProviderFactory,
+                                                    authService: authService)
+            endpoints[name] = endpointConfig
+        }
+
+        return endpoints
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/API/Sources/AWSAPIPlugin/Configuration/AWSAPICategoryPluginConfiguration.swift:199]: great question. A lot of the other categories will throw a `PluginError.configuration` error when it cannot find the category to configure with. Configuring **AWSAPIPlugin** should be the same, so instead of returning an empty endpoints, it should throw PluginError. I'll update the code

In the future, **AWSAPIPlugin** will support not just the endpoint set up from the data category but endpoints from other categories, such as `API` for REST/API Gateway endpoints. Our logic should then map to endpoints from these two places, `data` and `api` category, and if at the end of it `endpoints` is still empty then we should throw the PluginError.

File: AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+extension AWSPinpointAnalyticsPlugin {
+    public struct Options {
+        public static let defaultAutoFlushEventsInterval: TimeInterval = 60
+        public static let defaultTrackAppSession = true
+        public static let defaultAutoSessionTrackingInterval: TimeInterval = {
```
[the remark this answers is not in the document]
Comment (ruisebas) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift:14]: I _think_ they need to be public in order to be used in the public initializer

File: Amplify/Core/Configuration/Internal/AmplifyConfigurationInitialization.swift (hunk 3 above)
Comment (lawmicha) [reply Amplify/Core/Configuration/Internal/AmplifyConfigurationInitialization.swift:133]: renamed some of the internal method names as well, `decodeAmplifyConfiguration` -> `decodeAmplifyOutputsData`

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/Base/TestConfigHelper.swift [hunk 7]
```diff
@@ -6,16 +6,25 @@
...
+        ProcessInfo.processInfo.arguments.contains("GEN2")
+    }
+
     static func retrieveAmplifyConfiguration(forResource: String) throws -> AmplifyConfiguration {
 
         let data = try retrieve(forResource: forResource)
         return try AmplifyConfiguration.decodeAmplifyConfiguration(from: data)
     }
 
+    static func retrieveAmplifyConfigurationV2(forResource: String) throws -> AmplifyConfigurationV2 {
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/Base/TestConfigHelper.swift:23]: this should be `AmplifyOutputsData`, i must have written this before we did a switch to AmplifyOutputsData and forgot to update the integ tests since they don't build with main package

File: AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift:14]: yup, just tried to remove `public`, @ruisebas is right. I think I prefer these not to be public though. To achieve this, we'll have to duplicate storing these values twice, once in a property and again as the default value for the initializer parameters.

File: AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/Configuration/AWSPinpointAnalyticsPluginConfiguration.swift [hunk 8]
```diff
@@ -20,25 +20,14 @@ public struct AWSPinpointAnalyticsPluginConfiguration {
...
 
-    static let defaultAutoFlushEventsInterval = 60
-    static let defaultTrackAppSession = true
-    static let defaultAutoSessionTrackingInterval: Int = {
-    #if os(macOS)
-        .max
-    #else
-        5
-    #endif
-    }()
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/Configuration/AWSPinpointAnalyticsPluginConfiguration.swift:31]: I've turn these back to internal

File: AmplifyPlugins/Analytics/Tests/AnalyticsHostApp/AWSPinpointAnalyticsPluginIntegrationTests/AWSPinpointAnalyticsPluginIntegrationTests.swift [hunk 9]
```diff
@@ -18,14 +18,25 @@ import Network
 class AWSPinpointAnalyticsPluginIntergrationTests: XCTestCase {
 
     static let amplifyConfiguration = "testconfiguration/AWSPinpointAnalyticsPluginIntegrationTests-amplifyconfiguration"
+    static let amplifyOutputs = "testconfiguration/AWSPinpointAnalyticsPluginIntegrationTests-amplify_outputs"
     static let analyticsPluginKey = "awsPinpointAnalyticsPlugin"
     
+    var useGen2Configuration: Bool {
+        ProcessInfo.processInfo.arguments.contains("GEN2")
+    }
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Tests/AnalyticsHostApp/AWSPinpointAnalyticsPluginIntegrationTests/AWSPinpointAnalyticsPluginIntegrationTests.swift:26]: i think that is only used by the unit test targets. the host apps consume the amplify products and `AmplifyTestCommon` is not one of them, unless I missed something here

[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Tests/AnalyticsHostApp/AWSPinpointAnalyticsPluginIntegrationTests/AWSPinpointAnalyticsPluginIntegrationTests.swift:26]: Keeping this "GEN2" check within the test seems reasonable since that's where the argument for the test plan is also configured. If we keep it in a common place, then of course there's some less flexibility because all test plans have to use the same argument, which is the case now but may not be the case in the future or all test targets across the categories

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Helpers/ConfigurationHelper.swift [hunk 10]
```diff
@@ -71,7 +75,49 @@ struct ConfigurationHelper {
...
+        // parse `verificationMechanisms`
+        let verificationMechanisms: [UserPoolConfigurationData.VerificationMechanism] = config
+            .userVerificationTypes?
+            .compactMap { .init(from: $0) } ?? []
+
+        return UserPoolConfigurationData(poolId: config.userPoolId,
+                                         clientId: config.userPoolClientId,
+                                         region: config.awsRegion,
+                                         endpoint: nil, // Gen2 does not support this field
+                                         clientSecret: nil, // Gen2 does not support this field
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Helpers/ConfigurationHelper.swift:113]: The library has no way to detect migration customers, it just reads the `amplify_outputs.json` file. I think you're referring to a callout specifically on the migration docs. I believe at that point we may include a schema update to allow customers to set`clientSecret` to enable this feature and for the library to then map it to this internal config property. This first PR is more targeted towards the happy path of Gen2 experience

File: AmplifyPlugins/Auth/Tests/AuthHostApp/AuthIntegrationTests/AuthGen2IntegrationTests.xctestplan [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  "defaultOptions" : {
+    "commandLineArgumentEntries" : [
+      {
+        "argument" : "GEN2"
+      }
+    ]
+  },
+  "testTargets" : [
+    {
+      "skippedTests" : [
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Auth/Tests/AuthHostApp/AuthIntegrationTests/AuthGen2IntegrationTests.xctestplan:20]: these tests are incrementally enabled as we work through provisioning and testing against the Gen2 backend. The test plan itself acts as a way for us to inject the GEN2 argument to pivot on configuration lifecycle and enable select tests that we have ran already

File: AmplifyPlugins/API/Sources/AWSAPIPlugin/Configuration/AWSAPICategoryPluginConfiguration.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/API/Sources/AWSAPIPlugin/Configuration/AWSAPICategoryPluginConfiguration.swift:205]: Updated

File: AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift]: Updated

File: AmplifyPlugins/Auth/Tests/AuthHostApp/AuthIntegrationTests/SignUpTests/AuthSignUpTests.swift [hunk 12]
```diff
@@ -39,14 +37,13 @@ class AuthSignUpTests: AWSAuthBaseTest {
         for _ in 0..<signUpExpectation.expectedFulfillmentCount {
 
             Task {
-                let username = "integTest\(UUID().uuidString)"
-                let password = "P123@\(UUID().uuidString)"
+                let username = useGen2Configuration ? "test-\(UUID().uuidString)@amazon.com" : "integTest\(UUID().uuidString)"
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Auth/Tests/AuthHostApp/AuthIntegrationTests/SignUpTests/AuthSignUpTests.swift:40]: as discussed, we'll use username as the username with the CDK override. I'm deploying another auth backend and will test and revert some of these code changes

[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Auth/Tests/AuthHostApp/AuthIntegrationTests/SignUpTests/AuthSignUpTests.swift]: Updated configuration for Gen2 to be provisioned like Gen1

File: AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        }()
+
+        public let autoFlushEventsInterval: TimeInterval
+        public let trackAppSessions: Bool
+        public let autoSessionTrackingInterval: TimeInterval
+
+        #if os(macOS)
+        public init(autoFlushEventsInterval: TimeInterval = 60,
+                    trackAppSessions: Bool = true,
+                    autoSessionTrackingInterval: TimeInterval = .infinity) {
```
Comment (ruisebas) [inline AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift:29]: On a second though, are we sure we want to expose this flag, and if so with this name? 

It's usage it very unintuitive. When the app goes to the background, we set the current session to "paused" and this value is meant to be how much time it has to pass until it effectively becomes "stopped" if the app is still in the background (i.e. if the apps comes back to the foreground before this time interval, then we just "resume" the session instead of starting a new one; if more time has passed we start a new session).

But it's also not completely honoured, because we also keep track on how much the OS lets the app run in the background.
So if a customer sets this to 60 seconds, we will still stop the session once the OS stops background execution (which is around 5 seconds IIRC).

Also this not used on macOS at all. It's set to `.max` (or  `.infinity`) because we originally figured there was no limit on how long an app can be "backgrounded" in a desktop system, but eventually we decided to not pause sessions in macOS as the app still runs.

tl;dr: I don't think this value should be publicly exposed.

**Review by ruisebas — CHANGES_REQUESTED** [review]

File: AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift (hunk 13 above)
Comment (lawmicha) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift:29]: I've just been focused on parity in translating the capabilities in old to the new DX (config + options) so I didn't consider whether this option should even exist, thanks for the detailed explanation!

I'm inclined to remove this. we can always add it in later with better documentation on the usage, and clearer naming. I doubt anyone is actually overriding this in the old configuration file due to how unintuitive it is. If they set this to 1 second then a session is "stopped" almost immediately after the app goes to background. I think the use case is to "create a new session immediately if the app is moved to the background".  When setting this to a large value, the OS stops background execution without any guarantees, as you've said, so the value is not completely honoured.

Comment (lawmicha) [reply AmplifyPlugins/Analytics/Sources/AWSPinpointAnalyticsPlugin/AWSPinpointAnalyticsPlugin+Options.swift]: Removing `autoSessionTrackingInterval` in this PR, commit https://github.com/aws-amplify/amplify-swift/pull/3567/commits/9cbc295b6a8f94c33a989c09cf9910460a0da9c2

File: AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginHostApp/AWSCloudWatchLoggingPluginIntegrationTests/AWSCloudWatchLoggingPluginIntegrationTests.swift [hunk 14]
```diff
@@ -30,14 +35,21 @@ class AWSCloudWatchLoggingPluginIntergrationTests: XCTestCase {
...
+            }
+
             try await Task.sleep(seconds: 5)
         } catch {
             XCTFail("Failed to initialize and configure Amplify: \(error)")
         }
-        XCTAssertNotNil(Amplify.Auth.plugin)
-        XCTAssertTrue(Amplify.Auth.isConfigured)
+        //XCTAssertNotNil(Amplify.Auth.plugin)
+        //XCTAssertTrue(Amplify.Auth.isConfigured)
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Logging/Tests/AWSCloudWatchLoggingPluginHostApp/AWSCloudWatchLoggingPluginIntegrationTests/AWSCloudWatchLoggingPluginIntegrationTests.swift:52]: Yes, don't recall what happened here

File: AmplifyPlugins/Storage/Tests/StorageHostApp/AWSS3StoragePluginIntegrationTests/AWSS3StoragePluginAccessLevelTests.swift [hunk 15]
```diff
@@ -49,16 +51,25 @@ class AWSS3StoragePluginAccessLevelTests: AWSS3StoragePluginTestBase {
...
+
         let accessLevel: StorageAccessLevel = .guest
 
         do {
             logger.debug("Sign In")
             let result = try await Amplify.Auth.signIn(username: username, password: password)
             XCTAssertTrue(result.isSignedIn)
             let currentUser = try await Amplify.Auth.getCurrentUser()
-            XCTAssertEqual(username, currentUser.username)
+            //XCTAssertEqual(username, currentUser.username)
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Storage/Tests/StorageHostApp/AWSS3StoragePluginIntegrationTests/AWSS3StoragePluginAccessLevelTests.swift:72]: This is similar to Auth integration tests. Since updating the Auth backend to use Username as username for sign up, we can now uncomment this. (using email as username would fail the username check here.

File: AmplifyPlugins/Storage/Tests/StorageHostApp/AWSS3StoragePluginIntegrationTests/AWSS3StoragePluginAccessLevelTests.swift [hunk 16]
```diff
@@ -101,14 +112,21 @@ class AWSS3StoragePluginAccessLevelTests: AWSS3StoragePluginTestBase {
...
+        } else {
+            username = AWSS3StoragePluginTestBase.user1.lowercased()
+            password = AWSS3StoragePluginTestBase.password
+        }
         logger.debug("Signing in as user1")
         let result = try await Amplify.Auth.signIn(username: username, password: password)
         XCTAssertTrue(result.isSignedIn)
         let currentUser = try await Amplify.Auth.getCurrentUser()
-        XCTAssertEqual(username, currentUser.username)
+        // XCTAssertEqual(username, currentUser.username)
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply AmplifyPlugins/Storage/Tests/StorageHostApp/AWSS3StoragePluginIntegrationTests/AWSS3StoragePluginAccessLevelTests.swift]: updated, uncommented since provisioning steps have been updated

File: AmplifyPlugins/Predictions/AWSPredictionsPlugin/AWSPredictionsPlugin+Configure.swift [hunk 17]
```diff
@@ -19,21 +19,26 @@ extension AWSPredictionsPlugin {
...
-        guard let jsonValueConfiguration = configuration as? JSONValue else {
+        let predictionsConfiguration: PredictionsPluginConfiguration
+        if configuration is AmplifyOutputsData {
             throw PluginError.pluginConfigurationError(
-                PluginErrorMessage.decodeConfigurationError.errorDescription,
-                PluginErrorMessage.decodeConfigurationError.recoverySuggestion
+                PluginErrorMessage.amplifyOutputsConfigurationNotSupportedError.errorDescription,
+                PluginErrorMessage.amplifyOutputsConfigurationNotSupportedError.recoverySuggestion
+            )
+        } else {
```
Comment (phantumcode) [inline AmplifyPlugins/Predictions/AWSPredictionsPlugin/AWSPredictionsPlugin+Configure.swift:28]: nit: convert to `else if let` and `else` throws exception

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePlugin+Configure.swift [hunk 18]
```diff
@@ -18,35 +18,57 @@ extension AWSS3StoragePlugin {
...
     public func configure(using configuration: Any?) throws {
-        guard let config = configuration as? JSONValue else {
-            throw PluginError.pluginConfigurationError(PluginErrorConstants.decodeConfigurationError.errorDescription,
-                                                       PluginErrorConstants.decodeConfigurationError.recoverySuggestion)
-        }
+        let region: () throws -> String
+        let bucket: () throws -> String
+        var defaultAccessLevel: () throws -> StorageAccessLevel = { .guest }
+        if let config = configuration as? AmplifyOutputsData {
+            guard let storage = config.storage else {
```
Comment (phantumcode) [inline AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePlugin+Configure.swift:31]: there's a lot of nesting here, can this get refactored into smaller functions?

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Support/Constants/PluginErrorConstants.swift [hunk 19]
```diff
@@ -19,6 +19,10 @@ struct PluginErrorConstants {
         "Configuration was not a dictionary literal",
         "Make sure the value for the plugin is a dictionary literal with keys 'Bucket' and 'Region'")
 
+    static let missingStorageCategoryConfiguration: PluginErrorString = (
+        "Plugin is missing `Storage` category in configuration.",
+        "Add the `Storage` section to the plugin.")
```
Comment (phantumcode) [inline AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/Support/Constants/PluginErrorConstants.swift:24]: should be "Add the `Storage` section to the configuration"

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+public struct AWSS3StoragePluginOptions {
+    public let defaultAccessLevel: StorageAccessLevel
```
Comment (phantumcode) [inline AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift:12]: can you mark `deafultAccessLevel` as deprecated?

File: Amplify/Core/Configuration/AmplifyConfiguration.swift [hunk 21]
```diff
@@ -175,7 +175,7 @@ extension Amplify {
 
     /// Notifies all hub channels that Amplify is configured, in case any plugins need to be notified of the end of the
     /// configuration phase (e.g., to set up cross-channel dependencies)
-    private static func notifyAllHubChannels() {
+    static func notifyAllHubChannels() {
```
[the remark this answers is not in the document]
Comment (lawmicha) [reply Amplify/Core/Configuration/AmplifyConfiguration.swift:178]: AmplifyOutput's configuration methods also use this method and is in a separate file, which is why it was made `internal`

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift (hunk 20 above)
Comment (lawmicha) [reply AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift:12]: sorry, this options struct should not be added since `defaultAccessLevel` is being deprecated as part of the storage path PR, which leaves this options struct empty. we can introduce the options struct pattern once we do have some client configuration values in the future.

File: AmplifyPlugins/Predictions/AWSPredictionsPlugin/AWSPredictionsPlugin+Configure.swift (hunk 17 above)
Comment (lawmicha) [reply AmplifyPlugins/Predictions/AWSPredictionsPlugin/AWSPredictionsPlugin+Configure.swift]: Updating here, and all other plugins as well

File: AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift (hunk 20 above)
Comment (phantumcode) [reply AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift]: Can you remove this change/file then? It still shows the in the latest version

Comment (lawmicha) [reply AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift]: I've removed `AWSS3StoragePluginOptions.swift` file, can you check again? might be a delay from Github

Comment (lawmicha) [reply AmplifyPlugins/Storage/Sources/AWSS3StoragePlugin/AWSS3StoragePluginOptions.swift]: storage source files image

**Review by thisisabhash — APPROVED** [review]
Geo/Core - Approved

**Review by ruisebas — DISMISSED** [review]
Analytics/Push: Approved

**Review by phantumcode — APPROVED** [review]

**Comment by 5d** [comment]
API: Approved

**Review by 5d — DISMISSED** [review]

**Review by thisisabhash — APPROVED** [review]
